File: samples/ObjC Collection View Inbox/watchkitapp Extension/EventsController.m
/* 
 * Licensed Materials - Property of IBM 
 * 
 * 5725E28, 5725I03 
 * 
 * © Copyright IBM Corp. 2017, 2018
 * US Government Users Restricted Rights - Use, duplication or disclosure restricted by GSA ADP Schedule Contract with IBM Corp. 
 */


#import "EventsController.h"
#import <IBMMobilePushWatch/IBMMobilePushWatch.h>

@interface EventsController()
@property NSMutableArray * listeners;
@property NSTimer * queueEventTimer;
@property NSTimer * sendEventTimer;
@property NSTimer * sendQueueTimer;
@end

@implementation EventsController

-(IBAction)sendEvent:(id)sender
{
    [self.sendEventStatus setText:@"Sending"];
    [self.sendEventStatus setTextColor:UIColor.whiteColor];
    MCEEvent * event = [[MCEEvent alloc] initWithName: @"watch" type: @"watch" timestamp:nil attributes:@{ @"immediate": @YES }];
    [[MCEEventService sharedInstance] addEvent:event immediate:TRUE];
}

-(IBAction)queueEvent:(id)sender
{
    [self.queueEventStatus setText:@"Queued"];
    [self.queueEventStatus setTextColor:UIColor.whiteColor];
    MCEEvent * event = [[MCEEvent alloc] initWithName: @"watch" type:@"watch" timestamp:nil attributes:@{ @"immediate": @NO }];
    [[MCEEventService sharedInstance] addEvent:event immediate:FALSE];
}

-(IBAction)sendQueue:(id)sender
{
    [self.sendQueueStatus setText:@"Sending"];
    [self.sendQueueStatus setTextColor:UIColor.whiteColor];
    [[MCEEventService sharedInstance] sendEvents];
}

- (void)didDeactivate
{
    [super didDeactivate];
    for (id listener in self.listeners) {
        [NSNotificationCenter.defaultCenter removeObserver:listener];
    }
}

-(void)willActivate
{
    [super willActivate];
    
    [self.listeners addObject: [NSNotificationCenter.defaultCenter addObserverForName:MCEEventSuccess object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {

        for(MCEEvent * event in note.userInfo[@"events"])
        {
            if([event.type isEqual: @"watch"] && [event.name isEqual: @"watch"])
            {
                [self.sendQueueStatus setText:@"Received"];
                [self.sendQueueStatus setTextColor:UIColor.greenColor];
                if(self.sendQueueTimer)
                {
                    [self.sendQueueTimer invalidate];
                }
                self.sendQueueTimer = [NSTimer scheduledTimerWithTimeInterval:5 repeats:NO block:^(NSTimer * _Nonnull timer) {
                    [self.sendQueueStatus setText:@"Idle"];
                    [self.sendQueueStatus setTextColor:UIColor.lightGrayColor];
                    self.sendQueueTimer = nil;
                }];

                if([event.attributes[@"immediate"] boolValue])
                {
                    [self.sendEventStatus setText:@"Received"];
                    [self.sendEventStatus setTextColor:UIColor.greenColor];
                    if(self.sendEventTimer)
                    {
                        [self.sendEventTimer invalidate];
                    }
                    self.sendEventTimer = [NSTimer scheduledTimerWithTimeInterval:5 repeats:NO block:^(NSTimer * _Nonnull timer) {
                        [self.sendEventStatus setText:@"Idle"];
                        [self.sendEventStatus setTextColor:UIColor.lightGrayColor];
                        self.sendEventTimer = nil;
                    }];
                }
                else
                {
                    [self.queueEventStatus setText:@"Received"];
                    [self.queueEventStatus setTextColor:UIColor.greenColor];
                    if(self.queueEventTimer)
                    {
                        [self.queueEventTimer invalidate];
                    }
                    self.queueEventTimer = [NSTimer scheduledTimerWithTimeInterval:5 repeats:NO block:^(NSTimer * _Nonnull timer) {
                        [self.queueEventStatus setText:@"Idle"];
                        [self.queueEventStatus setTextColor:UIColor.lightGrayColor];
                        self.queueEventTimer = nil;
                    }];
                }
            }
        }
        
    }]];
    
    [self.listeners addObject: [NSNotificationCenter.defaultCenter addObserverForName:MCEEventFailure object:nil queue:NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {

        for(MCEEvent * event in note.userInfo[@"events"])
        {
            if([event.type isEqual: @"watch"] && [event.name isEqual: @"watch"])
            {
                [self.sendQueueStatus setText:@"Error"];
                [self.sendQueueStatus setTextColor:UIColor.redColor];
                if(self.sendQueueTimer)
                {
                    [self.sendQueueTimer invalidate];
                }
                if([event.attributes[@"immediate"] boolValue])
                {
                    [self.sendEventStatus setText:@"Error"];
                    [self.sendEventStatus setTextColor:UIColor.redColor];
                    if(self.sendEventTimer)
                    {
                        [self.sendEventTimer invalidate];
                    }
                }
                else
                {
                    [self.queueEventStatus setText:@"Error"];
                    [self.queueEventStatus setTextColor:UIColor.redColor];
                    if(self.queueEventTimer)
                    {
                        [self.queueEventTimer invalidate];
                    }
                }
            }
        }
    }]];
}

- (void)awakeWithContext:(id)context
{
    [super awakeWithContext:context];
    self.listeners = [NSMutableArray array];
}

@end

File: samples/Swift Automatic Sample/SampleWatch Extension/EventsController.swift
/* IBM Confidential
 * OCO Source Materials
 * 5725E28, 5725S01, 5725I03
 * © Copyright IBM Corp. 2017, 2017
 *
 * The source code for this program is not published or otherwise
 * divested of its trade secrets, irrespective of what has been
 * deposited with the U.S. Copyright Office.
 */

import Foundation
import WatchKit

class EventsController: WKInterfaceController
{
    @IBOutlet weak var sendEventStatus: WKInterfaceLabel?
    @IBOutlet weak var queueEventStatus: WKInterfaceLabel?
    @IBOutlet weak var sendQueueStatus: WKInterfaceLabel?
    var listeners = [NSObjectProtocol]()
    var queueEventTimer: Timer?
    var sendEventTimer: Timer?
    var sendQueueTimer: Timer?
    
    @IBAction func sendEvent(sender: Any) {
        sendEventStatus?.setText("Sending")
        sendEventStatus?.setTextColor(.white)
        let event = MCEEvent(name: "watch", type: "watch", timestamp: nil, attributes: ["immediate": true])
        MCEEventService.sharedInstance().add(event, immediate: true)
    }
    
    @IBAction func queueEvent(sender: Any) {
        queueEventStatus?.setText("Queued")
        queueEventStatus?.setTextColor(.white)
        let event = MCEEvent(name: "watch", type: "watch", timestamp: nil, attributes: ["immediate": false])
        MCEEventService.sharedInstance().add(event, immediate: false)
    }
    
    @IBAction func sendQueue(sender: Any) {
        sendQueueStatus?.setText("Sending")
        sendQueueStatus?.setTextColor(.white)
        MCEEventService.sharedInstance().sendEvents()
    }
    
    override func didDeactivate() {
        super.didDeactivate()
        for listener in listeners {
            NotificationCenter.default.removeObserver(listener)
        }
    }
    
    override func willActivate() {
        super.willActivate()
        listeners.append(NotificationCenter.default.addObserver(forName: NSNotification.Name("MCEEventSuccess"), object: nil, queue: OperationQueue.main, using: { (note) in
            for event in note.userInfo?["events"] as! [MCEEvent]
            {
                if event.type == "watch" && event.name == "watch"
                {
                    self.sendQueueStatus?.setText("Received")
                    self.sendQueueStatus?.setTextColor(.green)
                    if let sendQueueTimer = self.sendQueueTimer
                    {
                        sendQueueTimer.invalidate()
                    }
                    self.sendQueueTimer = Timer.scheduledTimer(withTimeInterval: 5, repeats: false, block: { (timer) in
                        self.sendQueueStatus?.setText("Idle")
                        self.sendQueueStatus?.setTextColor(.lightGray)
                        self.sendQueueTimer = nil
                    })
                    
                    if event.attributes["immediate"] as! Bool
                    {
                        self.sendEventStatus?.setText("Received")
                        self.sendEventStatus?.setTextColor(.green)
                        if let sendEventTimer = self.sendEventTimer
                        {
                            sendEventTimer.invalidate()
                        }
                        self.sendEventTimer = Timer.scheduledTimer(withTimeInterval: 5, repeats: false, block: { (timer) in
                            self.sendEventStatus?.setText("Idle")
                            self.sendEventStatus?.setTextColor(.lightGray)
                            self.sendEventTimer = nil
                        })
                    }
                    else
                    {
                        self.queueEventStatus?.setText("Received")
                        self.queueEventStatus?.setTextColor(.green)
                        if let queueEventTimer = self.queueEventTimer
                        {
                            queueEventTimer.invalidate()
                        }
                        self.queueEventTimer = Timer.scheduledTimer(withTimeInterval: 5, repeats: false, block: { (timer) in
                            self.queueEventStatus?.setText("Idle")
                            self.queueEventStatus?.setTextColor(.lightGray)
                            self.queueEventTimer = nil
                        })
                    }
                }
            }
        }))
        
        listeners.append(NotificationCenter.default.addObserver(forName: NSNotification.Name("MCEEventFailure"), object: nil, queue: OperationQueue.main, using: { (note) in
            
            for event in note.userInfo?["events"] as! [MCEEvent]
            {
                if event.type == "watch" && event.name == "watch"
                {
                    self.sendQueueStatus?.setText("Error")
                    self.sendQueueStatus?.setTextColor(.red)
                    if let sendQueueTimer = self.sendQueueTimer
                    {
                        sendQueueTimer.invalidate()
                    }
                    
                    if event.attributes["immediate"] as! Bool
                    {
                        self.sendEventStatus?.setText("Error")
                        self.sendEventStatus?.setTextColor(.red)
                        if let sendEventTimer = self.sendEventTimer
                        {
                            sendEventTimer.invalidate()
                        }
                    }
                    else
                    {
                        self.queueEventStatus?.setText("Error")
                        self.queueEventStatus?.setTextColor(.red)
                        if let queueEventTimer = self.queueEventTimer
                        {
                            queueEventTimer.invalidate()
                        }
                    }
                }
            }
        }))
    }
}
